### ISSUE TITLE
`random.split` should support splitting starting at a particular index, such as `step` number

### ISSUE THREAD

**Issue by reinerp** [google/jax#7763]
Background: during training, we can deterministically derive the PRNG seed from the current step number, by `seed=hash(step_number)`. This can sometimes be convenient because it avoids the need to checkpoint the PRNG seed, as the step number is all the state that is needed.

The computation `seed=hash(step_number)` can be implemented, albeit _inefficiently_, in JAX:

```
key_this_step = jax.random.split(global_key, step_number + 1)[step number]
```

Sadly, this is inefficient in that it materializes an intermediate array of length `step_number + 1`, even though we only want to access the last element of that array. Given that all of the elements of `jax.random.split` are computed in parallel anyway, it would be nice to directly have access to the `n`th split of `jax.random.split` rahter than having to produce all of them and selecting one.

One way to integrate this into the API is to add a parameter `slice=None` to `jax.random.split`, whose behavior is defined by ensuring that the following equivalence holds:

```
jax.random.split(key, slice=(lo, hi)) == jax.random.split(key, num)[lo:hi]
```

I note that there are other requests for extensions to `jax.random.split`, such as https://github.com/google/jax/issues/4013. Another alternative, which seems flexible enough to serve everyone's need, would be to directly expose the underlying hash function as a function `jax.random.hash(key, index)`, defined so that the following equivalence holds:

```
jax.random.split(key, num) == jax.random.hash(key, jax.lax.iota(jnp.uint32, num))
```

**Comment by jakevdp** [comment]
Thanks for the suggestion! This is a really good time to bring this up, because @froystig has just finished a refactoring of the PRNG code so that users can override the API, including `_split`. This means that if we are planning to make any changes to what options `_split` might accept, now would be the time (before downstream libraries implement their own random generators with their own `split` methods).

I wonder if @froystig has thoughts about this request, and the one in #4013?

**Comment by jakevdp** [comment]
Also, one quick comment: you should be aware that `random.split(N - 1)` does not return values that are identical to the first `N - 1` entries of `random.split(N)`:
```python
from jax import random
import jax.numpy as jnp
key = random.PRNGKey(0)

print(random.split(key, 5))
# [[1524306142 2595015335]
#  [1836460763  990488084]
#  [1416732029 1887795613]
#  [1078027127 1191019179]
#  [3297765038 3069809391]]

print(random.split(key, 4))
# [[2285895361 1501764800]
#  [1518642379 4090693311]
#  [ 433833334 4221794875]
#  [ 839183663 3740430601]]
```
It seems like your approach might implicitly assume that.

**Comment by reinerp** [comment]
Thanks for pointing that out, I'm glad you caught that issue about `random.split(N-1)` vs `random.split(N)`!

I hadn't realized that `random.threefry_2x32` is publicly visible. Indeed, that does what I'm looking for!

**Comment by froystig** [comment]
Would `random.fold_in(key, step)` work?

By the way, `random.threefry_2x32` doesn't want to be visible any longer, and issues a deprecation warning:

https://github.com/google/jax/blob/5f46206898a9052a140b51da6a1ed24731a20d47/jax/_src/random.py#L1449-L1452

As of #6899, PRNG implementations are swappable, so one shouldn't rely on the default PRNG internals (e.g. threefry) or mix uses of `prng.threefry_2x32` in with generic RNG operations.

**Comment by reinerp** [comment]
`random.fold_in(key, step)` looks great, thanks for pointing me to i!

**Comment by jekbradbury** [comment]
Yeah, I think that's the API intended for this. One recent question that was raised is whether we could also make them give the same results (i.e. make `split(key) == (fold_in(key, 0), fold_in(key, 1))`).

**Comment by froystig** [comment]
> One recent question that was raised is whether we could also make them give the same results (i.e. make `split(key) == (fold_in(key, 0), fold_in(key, 1))`).

Some discussion to this end is in #7708.
